Hive registration initial commit

zliu41 opened the pull request:
This is the initial commit for Hive registration. It mainly contains interfaces, abstract classes, base implementations, as well as the main class `HiveRegister`.

- `HiveRegister` performs Hive registration given a `HiveSpec`, which specifies which path, which db, which table, etc., to register.  There are several extensions to `HiveSpec`, including `HiveSpecWithPredicates`, `HiveSpecWithPreActivities` and `HiveSpecWithPostActivities`. A base implementation `SimpleHiveSpec` is included.

- A `HiveSpec` is generated by a `HiveRegistrationPolicy` given a `Path`. A base implementation `HiveRegistrationPolicyBase` is included.

- To do Hive registration, a `StorageDescriptor` and some schema properties are often required. These are managed by `HiveStorageDescriptorManager` and `HiveSchemaManager`.

zliu41 commented:
@sahilTakiar @ydai1124 can you review?

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,302 @@
+      public Void call() throws Exception {
+
+        if (spec instanceof HiveSpecWithPredicates && !evaluatePredicates((HiveSpecWithPredicates) spec)) {
+          log.info("Skipping " + spec + " since predicates return false");
```
I think you need a `return` after this line.

zliu41 replied:
fixed

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,302 @@
+
+        if (spec instanceof HiveSpecWithPreActivities) {
+          for (Activity<HiveMetaStoreClient> activity : ((HiveSpecWithPreActivities) spec).getActivities()) {
+            activity.execute(HiveRegister.this.client);
```
Do we expect a requirement to execute these activities in parallel?

zliu41 replied:
Not really. This is for the use case of dropping the hourly partitions after registering the daily partition. I'm not even sure whether dropping multiple partitions from the same table can be parallelized or not, but even if it can it likely won't bring any significant performance gain.

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,302 @@
+    try {
+      client.createDatabase(db);
+    } catch (AlreadyExistsException e) {
+      log.warn("Hive database " + dbName + " already exists");
```
We probably don't need these since (besides the initial run), this will always be printed out.

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,302 @@
+      log.warn("Hive database " + dbName + " already exists");
+    } catch (TException e) {
+      log.error("Unable to create Hive database " + dbName);
+      throw Throwables.propagate(e);
```
Why do we need to log the error and re-throw it? I would either, wrap it in a checked exception like `CannotCreateHiveDBError` and re-throw it, or just `throw Throwables.propagate(e)`

zliu41 replied:
This is just in case the exception message is not clear enough. Adding a log here to ensure printing a message that we can control.

sahilTakiar replied:
I generally think throwing a checked exception liked `CannotCreateHiveDBError` is a better way of adding a custom message to any exception thrown.

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,302 @@
+  private Table createTable(String dbName, String tableName, Optional<HivePartition> partition, StorageDescriptor sd) {
+    Table table = new Table();
+    table.setParameters(Maps.<String, String> newHashMap());
+    table.getParameters().put("EXTERNAL", Boolean.TRUE.toString());
```
Why don't you built the parameters in a `Map` first and then call `table.setParameters(map)`.

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,302 @@
+      table.setPartitionKeys(partition.get().getKeys());
+    }
+
+    Preconditions.checkState(this.hiveOwner.isPresent(), "Missing required property " + HiveRegProps.HIVE_OWNER);
```
`checkState` to `checkArgument`

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,302 @@
+    return table;
+  }
+
+  public void deregister(String dbName, String tableName) {
```
Method name sounds a little value, deregistering what exactly? Recommend changing to `deregisterTable` or `dropTable`. Similar comment for the overloaded methods.

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,302 @@
+  }
+
+  private void addPartition(Table table, HivePartition partition) throws TException {
+    Preconditions.checkArgument(table.getPartitionKeysSize() == partition.getValues().size());
```
Probably need to check if the `dbName` `tableName` and `serDe` of the given `table` and `partition` match.

zliu41 replied:
This method takes a `Table` and partition names and values. The actual `Partition` object is constructed in this method using the dbname, tablename and serde of the given `Table`.

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,302 @@
+    this.hiveDbRootDir = this.props.getDbRootDir();
+    this.hiveOwner = this.props.getHiveOwner();
+    this.client = new HiveMetaStoreClient(new HiveConf());
+    this.executor = MoreExecutors.listeningDecorator(
```
What's the point of using `MoreExecutors.listeningDecorator`?

zliu41 replied:
`register()` returns a `ListenableFuture` so that the caller of `register()` can add whatever callbacks they want.

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveSchemaManager.java:
```diff
@@ -0,0 +1,92 @@
+   * @param props A {@link State} object used to instantiate {@link HiveSchemaManager}.
+   * @return
+   */
+  public static HiveSchemaManager getInstance(String type, State props) {
```
Consider using a `[ServiceLoader](https://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html)` instead of this pattern. PR https://github.com/linkedin/gobblin/pull/627 has an example of how to use it. This way each implementation of `HiveSchemaManager` does not have to explicitly register itself with this class. It should also allow you to make this class an interface instead of an abstract class.

zliu41 replied:
Took a look at `ServiceLoader`, it seems to require that the concrete classes have zero-argument constructor. But `HiveSchemaManager` needs to be instantiated with a `State`.

sahilTakiar replied:
`HiveSchemaManager` can just have a `configure(State state)` method, and probably a `String getType()` method that returns either `nop` or `avro`. Functionality its the same, so its up to you.

zliu41 replied:
That's true. But if the user directly instantiates a `HiveSchemaManager` implementation using the constructor, he may forget to call `configure(state)`. So for me the current approach is preferred

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveNopSchemaManager.java:
```diff
@@ -0,0 +1,40 @@
+ *
+ * @author ziliu
+ */
+@Alpha
```
I think the `Alpha` annotations are typically meant just for public APIs (e.g. interfaces or abstract classes), although there is no harm in tagging concrete class too.

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,302 @@
+        }
+
+        if (spec instanceof HiveSpecWithPreActivities) {
+          for (Activity<HiveMetaStoreClient> activity : ((HiveSpecWithPreActivities) spec).getActivities()) {
```
Should the `Activity` take in a `HiveMetaStoreClient`? Maybe it should take in a `HiveRegister` instead? I'm guessing a common use case for pre and post activities would be to drop partitions? Since the `HiveRegister` class already has methods to drop partitions it may be best to use those, especially since those both are protected by locks. Providing unrestricted access to `HiveMetaStoreClient` may be dangerous since the registration is multi-threaded.

zliu41 replied:
great point

zliu41 commented:
@sahilTakiar updated 45d36e8

ydai1124 commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,351 @@
+   *
+   * @return a {@link ListenableFuture} for the process of registering the given {@link HiveSpec}.
+   */
+  public ListenableFuture<Void> register(final HiveSpec spec) throws IOException {
```
Is it possible to have multiple `HiveSpec`, e.g., both `HiveSpecWithPredicates` and `HiveSpecWithPreActivities`?

zliu41 replied:
A concrete `HiveSpec` implementation can implement both `HiveSpecWithPredicates` and `HiveSpecWithPreActivities`

ydai1124 commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,351 @@
+
+    Preconditions.checkState(this.hiveDbRootDir.isPresent(),
+        "Missing required property " + HiveRegProps.HIVE_DB_ROOT_DIR);
+    db.setLocationUri(this.hiveDbRootDir.get() + Path.SEPARATOR + dbName + HIVE_DB_EXTENSION);
```
Do you need to deal with multiple `Path.SEPARATOR` if one put one more separator to the end of `this.hiveDbRootDir.get()`? If that is the case, you can use `new Path(String, String).toString()` to deal with that.

ydai1124 commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,351 @@
+      client.createDatabase(db);
+      log.info("Created database " + dbName);
+    } catch (AlreadyExistsException e) {
+      // Database already exists. Nothing to do.
```
Can you add one more log if the db exists?

zliu41 replied:
This may flood the log. The db will only be created once, and after that every single registration of a partition will print this log msg.

ydai1124 commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,351 @@
+  }
+
+  public boolean existsPartition(String dbName, String tableName, List<String> partitionValues) {
+    Lock lock = locks.get(dbName + tableName);
```
Do you need to lock the whole table when checking the existence of some partitions?

zliu41 replied:
Instead of locking on a single `HiveMetaStoreClient` object, I probably need to use a pool of `HiveMetaStoreClient` objects, since `HiveMetaStoreClient` doesn't seem to support concurrency.

ydai1124 commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,351 @@
+  }
+
+  private void addPartition(Table table, HivePartition partition, Path partitionLocation) throws TException {
+    Preconditions.checkArgument(table.getPartitionKeysSize() == partition.getValues().size());
```
When could this mismatch happen?

zliu41 replied:
If someone creates a table that partitions on column A, and then try to add a partition whose partition key is A and B.

That being said I should probably compare the actual partition columns rather than the size.

ydai1124 commented on gobblin-hive-registration/src/main/java/gobblin/hive/avro/HiveAvroSchemaManager.java:
```diff
@@ -0,0 +1,84 @@
+    try {
+      for (FileStatus status : this.fs.listStatus(path)) {
+        if (status.isDir()) {
+          Schema schema = getSchema(status.getPath());
```
If there are multiple subdirs, are we assuming files under all subdirs have the same schema?

zliu41 replied:
Yes, it is not `HiveRegister`'s responsibility to verify that, because it is too expensive. If some user tries to register a path that contains different schemas, the registration will simply fail.

sahilTakiar commented on gobblin-hive-registration/build.gradle:
```diff
@@ -0,0 +1,82 @@
+    maxParallelForks = 1
+}
+
+clean {
```
Is this necessary? Same applies to the `artifacts` and `testJar` closures and the `tests` configuration.

zliu41 replied:
probably not. Don't remember where I copied this from.

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HivePartition.java:
```diff
@@ -0,0 +1,56 @@
+    for (int i = 0; i < keys.size(); i++) {
+      sb.append(keys.get(i).getName());
+      sb.append("=");
+      sb.append(values.get(i));
```
Probably want a `, ` in between each key value pair

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegProps.java:
```diff
@@ -0,0 +1,97 @@
+  @Getter
+  private final State tableProps;
+
+  public HiveRegProps(State props) {
```
Why not make this constructor take in two `State` objects, one `props` and the other `tableProps`. Seems cleaner the forcing the caller to put all table props into `hive.table.props`.

zliu41 replied:
yes I can add another constructor that takes two arguments.

sahilTakiar replied:
Why have the `HiveRegProps(State props)` constructor at all? It seems weird to force all values of `tablesProps` into a single props value. It's likely the invoker of this constructor will have to manually do this, and they won't realize they have to do this until they hit a bug.

zliu41 replied:
Because if we run it on Azkaban, then the program starts with a single property object from Azkaban. Someone has to separate tableProps from the property object somewhere, if not here then somewhere else. I don't see a reason why it should be somewhere else.

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/policy/HiveRegistrationPolicyBase.java:
```diff
@@ -0,0 +1,112 @@
+  public static final String HIVE_TABLE_REGEX = "hive.table.regex";
+
+  private static final Pattern NAME_PATTERN_1 = Pattern.compile("[a-z0-9][a-z0-9_]*");
+  private static final Pattern NAME_PATTERN_2 = Pattern.compile(".*[a-z_].*");
```
Make names of these variables more descriptive.

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/policy/HiveRegistrationPolicyBase.java:
```diff
@@ -0,0 +1,112 @@
+ * @author ziliu
+ */
+@Alpha
+public abstract class HiveRegistrationPolicyBase implements HiveRegistrationPolicy {
```
I would suggest splitting this into two classes one for `HiveRegisrationPolicyFromState` and `HiveRegistrationPolicyFromRegex`. The former checks the given `State` to see if the db and table are specified, the later uses a regex to find the db and table.

Is there a concrete use case for having a policy where the state is first checked, and then a regex is used?

zliu41 replied:
Yes, for example for tracking, the db name is just "tracking". The table name is a component in the path, which can be retrieved using a regex.

cwsteinbach commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,351 @@
+
+    Preconditions.checkArgument(this.hiveOwner.isPresent(), "Missing required property " + HiveRegProps.HIVE_OWNER);
+    table.setOwner(this.hiveOwner.get());
+    table.setTableType(TableType.EXTERNAL_TABLE.toString());
```
Why are you assuming that this is an external table?

zliu41 replied:
I guess this class is for the purpose of registering a specific path ingested by Gobblin in Hive. So it should be external table.

cwsteinbach commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,351 @@
+  }
+
+  public boolean existsTable(String dbName, String tableName) {
+    Lock lock = locks.get(dbName + tableName);
```
"dbName + tableName" is not a unique lock ID. For example the tables a.bc and ab.c both map to the same lock ID.

cwsteinbach replied:
I recommend creating a TableLock subclass which specifically takes dbName and tableName as input parameters in order to abstract out the creation of the lock ID.

zliu41 replied:
Instead of locking on a single `HiveMetaStoreClient` object, I probably need to use a pool of `HiveMetaStoreClient` objects, since `HiveMetaStoreClient` doesn't seem to support concurrency.

cwsteinbach commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,351 @@
+      log.info(String.format("Added partition %s to table %s with location %s", partition, table.getTableName(),
+          p.getSd().getLocation()));
+    } catch (AlreadyExistsException e) {
+      // Partition already exists. Nothing to do.
```
I think it's worth emitting a log message about this. Also, what if the partition that already exists has different values for the StorageDescriptor, location, etc?

zliu41 replied:
good point, will do alter_table and alter_partition if the new table or partition is different.

cwsteinbach commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,351 @@
+      client.dropPartition(dbName, tableName, partitionValues, false);
+      log.info("Dropped partition " + partitionValues + " in table " + tableName + " in db " + dbName);
+    } catch (NoSuchObjectException e) {
+      // Partition does not exist. Nothing to do
```
Worth logging a WARN message here.

zliu41 replied:
I'll change the name of this method to `dropPartitionIfExists`. Since it is normal that a partition doesn't exist, there shouldn't be a log warn.

zliu41 commented:
@sahilTakiar @ydai1124 @cwsteinbach updated 99035f7

Main changes:
- Use a pool of `HiveMetaStoreClient`s instead of a single `HiveMetaStoreClient`, because `HiveMetaStoreClient` doesn't seem to support concurrent requests.
- Added a striped lock class (`HiveStripedLocks`) as suggested by @cwsteinbach 
- When creating a table/partition, if the table/partition already exists, it checks whether the table or partition needs to be altered (`needToUpdateTable` and `needToUpdatePartition`). For example, update is needed if there's a schema change. If so it alters the table or partition.
- Addressed other comments.

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/policy/HiveRegistrationPolicyBase.java:
```diff
@@ -0,0 +1,143 @@
+      name = this.props.getProp(nameKey);
+    } else if (this.props.contains(regexKey)) {
+      String regex = this.props.getProp(regexKey);
+      name = Pattern.compile(regex).matcher(path.toString()).group();
```
Make a instance variable for the result of `Pattern.compile(regex)` since its will be expensive to compile a regex for every give `Path`

ydai1124 replied:
The `regex` here is only used for matching the string representation of `path`. But the `HIVE_DATABASE_REGEX`/ `HIVE_DATABASE_REGEX` is very generic naming. Can we rename it?

zliu41 replied:
changed to `dbNamePattern` and `tableNamePattern`.

sahilTakiar commented on gobblin-hive-registration/build.gradle:
```diff
@@ -0,0 +1,69 @@
+  testRuntime externalDependency.derby
+}
+
+task testJar(type: Jar, dependsOn: testClasses) {
```
Don't need this

sahilTakiar commented on gobblin-hive-registration/build.gradle:
```diff
@@ -0,0 +1,69 @@
+    from sourceSets.test.output
+}
+
+configurations {
```
Don't need this either.

sahilTakiar commented on gobblin-hive-registration/build.gradle:
```diff
@@ -0,0 +1,69 @@
+    compile externalDependency.avroMapredH1
+  }
+
+  testCompile externalDependency.calciteCore
```
Don't think you need any `testCompile` dependencies as there are no tests.

ydai1124 commented on gobblin-hive-registration/src/main/java/gobblin/hive/avro/HiveAvroSchemaManager.java:
```diff
@@ -0,0 +1,84 @@
+    Schema schema = getSchema(path);
+
+    Preconditions.checkNotNull(schema);
+    log.info("Schema is " + schema);
```
Can you add the `path` info to this log message, e.g., `Schema retrieved from <path> is ...`

zliu41 replied:
I'll actually remove this line since printing every table's schema is too much for driver log

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,413 @@
+  }
+
+  private boolean needToUpdateTable(Table existingTable, Table newTable) {
+    return !existingTable.equals(newTable);
```
How is this going to work? Looking at http://grepcode.com/file/repo1.maven.org/maven2/org.apache.hive/hive-metastore/1.0.1/org/apache/hadoop/hive/metastore/api/Table.java#Table.equals%28org.apache.hadoop.hive.metastore.api.Table%29 a `Table` is compared based on its create time, so I would guess the result of `createTableWithoutRegistering` will never be equal to `existingTable`

zliu41 replied:
My understanding is a Table's create time is not set by default, and will only be set if one explicitly does so. Otherwise it is 0.
Even if it is always automatically set, it shouldn't cause any problem if we always alter the table just to update the create time?

sahilTakiar replied:
Then `needToUpdateTable` will basically return true every time, because `existingTable` will have a valid created date (like 2015-02-08 00:00:01), but `newTable` will always have a created date of 0. So `alterTable` will be called every time, which won't cause any problems, but will be slightly less efficient.

What are the actual parameters we need to look for when deciding if the table needs to be altered? Can't there be a custom `Comparator` for comparing `Table` objects.

zliu41 replied:
I can set newTable's create time to be the same as existingTable's create time, then call equals().

sahilTakiar replied:
Yeah, that should work, although you will probably have to do this same for `lastAccessTime` and maybe even some other ones, there are some other fields in the `Table.equals` method, but I'm not sure what they all mean.

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveMetaStoreClientFactory.java:
```diff
@@ -0,0 +1,30 @@
+
+  @Override
+  public HiveMetaStoreClient makeObject() throws Exception {
+    return new HiveMetaStoreClient(new HiveConf());
```
When will all the `HiveMetaStoreClient`s be closed?

zliu41 replied:
good point

zliu41 replied:
Added a destroyObject method in `HiveMetaStoreClientFactory` which will take care of this.

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveMetaStoreClientFactory.java:
```diff
@@ -0,0 +1,30 @@
+public class HiveMetaStoreClientFactory extends BasePoolableObjectFactory<HiveMetaStoreClient> {
+
+  @Override
+  public HiveMetaStoreClient makeObject() throws Exception {
```
This should probably return `IMetaStoreClient` instead as according to the java docs that seems to be the public API:

https://hive.apache.org/javadocs/r1.0.1/api/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.html
https://hive.apache.org/javadocs/r1.0.1/api/org/apache/hadoop/hive/metastore/IMetaStoreClient.html

zliu41 replied:
agreed

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java:
```diff
@@ -0,0 +1,413 @@
+  private Table createTableWithoutRegistering(String dbName, String tableName, Optional<HivePartition> partition,
+      StorageDescriptor sd) {
+    Map<String, String> parameters = Maps.newHashMap();
+    parameters.put("EXTERNAL", Boolean.TRUE.toString());
```
Is this needed considering the `table.setTableType(TableType.EXTERNAL_TABLE.toString())` is also invoked.

zliu41 replied:
@cwsteinbach can chime in here but this is what the current code does.

ydai1124 commented on gobblin-hive-registration/src/main/java/gobblin/hive/policy/HiveRegistrationPolicyBase.java:
```diff
@@ -0,0 +1,143 @@
+      name = Pattern.compile(regex).matcher(path.toString()).group();
+    }
+
+    if (isNameValid(name)) {
```
Can we have one more log here?

zliu41 replied:
If the name is not valid, the exception message will be clear enough.

sahilTakiar commented on gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegProps.java:
```diff
@@ -0,0 +1,116 @@
+   * e.g., orc.compress=SNAPPY
+   */
+  public HiveRegProps(State props, State tableProps) {
+    this.tableProps = tableProps;
```
Missing `super(props)`

zliu41 commented:
@sahilTakiar @ydai1124 updated 7855bf7

sahilTakiar commented:
A few more minor comments, overall it looks good.

ydai1124 commented:
LGTM
